### PR TITLE
persist nsync.a across different platform builds

### PR DESCRIPTION
**Pull request by resec**
Persist nsync.a across different platform builds via copying them into gen folder, then later make can manage them (clean or something else). nsync.a is required when the built tensorflow static lib is linked by other libs, while currently the nsync.a is cleaned from each build.

### CODE REVIEW

**Comment by sb2nov** [comment]
Jenkins, test this please.

File: tensorflow/contrib/makefile/compile_nsync.sh [hunk 1]
```diff
@@ -302,7 +308,9 @@ done
 case "$target_platform" in
 ios)    nsync_platform_dir="$nsync_builds_dir/lipo.$target_platform.c++11"
         mkdir "$nsync_platform_dir"
```
Comment (drpngx) [inline tensorflow/contrib/makefile/compile_nsync.sh]: remove previous line?

**Review by drpngx — CHANGES_REQUESTED** [review]

**Comment by drpngx** [comment]
/CC @yifeif 

Can you describe a little more about the use case? I'm not sure I understand why this is needed.

**Comment by resec** [comment]
@drpngx when I am trying to build the another ios lib that is linking the Tensorflow static lib, the nsync.a is required as well for linking. If not provided, then below error would occur.

```
Showing Recent Messages
Ld /Users/resec/Library/Developer/Xcode/DerivedData/aabb-gflqvuqcsigceaevcvqprvmfrfol/Build/Products/Debug-iphonesimulator/aabb.app/aabb normal x86_64
    cd /project/aabb_ios_demo
    export IPHONEOS_DEPLOYMENT_TARGET=10.2
    export PATH="/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/usr/bin:/Applications/Xcode.app/Contents/Developer/usr/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin"
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator10.2.sdk -L/Users/resec/Library/Developer/Xcode/DerivedData/aabb-gflqvuqcsigceaevcvqprvmfrfol/Build/Products/Debug-iphonesimulator -F/Users/resec/Library/Developer/Xcode/DerivedData/aabb-gflqvuqcsigceaevcvqprvmfrfol/Build/Products/Debug-iphonesimulator -filelist /Users/resec/Library/Developer/Xcode/DerivedData/aabb-gflqvuqcsigceaevcvqprvmfrfol/Build/Intermediates/aabb.build/Debug-iphonesimulator/aabb.build/Objects-normal/x86_64/aabb.LinkFileList -Xlinker -rpath -Xlinker @executable_path/Frameworks -mios-simulator-version-min=10.2 -dead_strip -Xlinker -object_path_lto -Xlinker /Users/resec/Library/Developer/Xcode/DerivedData/aabb-gflqvuqcsigceaevcvqprvmfrfol/Build/Intermediates/aabb.build/Debug-iphonesimulator/aabb.build/Objects-normal/x86_64/aabb_lto.o -Xlinker -export_dynamic -Xlinker -no_deduplicate -Xlinker -objc_abi_version -Xlinker 2 -stdlib=libc++ -fobjc-arc -fobjc-link-runtime /project/agent/aabb/gen/sentencepiece/lib/libsentencepiece.a /project/agent/aabb/gen/protobuf_ios/lib/libprotobuf.a -force_load /project/agent/aabb/gen/lib/libtensorflow-core.a -Xlinker -sectcreate -Xlinker __TEXT -Xlinker __entitlements -Xlinker /Users/resec/Library/Developer/Xcode/DerivedData/aabb-gflqvuqcsigceaevcvqprvmfrfol/Build/Intermediates/aabb.build/Debug-iphonesimulator/aabb.build/aabb.app.xcent -lz -framework Accelerate -Xlinker -dependency_info -Xlinker /Users/resec/Library/Developer/Xcode/DerivedData/aabb-gflqvuqcsigceaevcvqprvmfrfol/Build/Intermediates/aabb.build/Debug-iphonesimulator/aabb.build/Objects-normal/x86_64/aabb_dependency_info.dat -o /Users/resec/Library/Developer/Xcode/DerivedData/aabb-gflqvuqcsigceaevcvqprvmfrfol/Build/Products/Debug-iphonesimulator/aabb.app/aabb

Undefined symbols for architecture x86_64:
  "nsync::nsync_mu_rlock(nsync::nsync_mu_s_*)", referenced from:
      tensorflow::ResourceMgr::DoLookup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::type_index, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, tensorflow::ResourceBase**) const in libtensorflow-core.a(resource_mgr.o)
  "nsync::nsync_mu_runlock(nsync::nsync_mu_s_*)", referenced from:
      tensorflow::ResourceMgr::DoLookup(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::type_index, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, tensorflow::ResourceBase**) const in libtensorflow-core.a(resource_mgr.o)
  "nsync::nsync_cv_wait(nsync::nsync_cv_s_*, nsync::nsync_mu_s_*)", referenced from:
      tensorflow::Notification::WaitForNotification() in libtensorflow-core.a(direct_session.o)
      tensorflow::Notification::WaitForNotification() in libtensorflow-core.a(graph_runner.o)
      tensorflow::Notification::WaitForNotification() in libtensorflow-core.a(cancellation.o)
      tensorflow::Notification::WaitForNotification() in libtensorflow-core.a(op_kernel.o)
      tensorflow::Notification::WaitForNotification() in libtensorflow-core.a(reader_base.o)
      tensorflow::Notification::WaitForNotification() in libtensorflow-core.a(rendezvous.o)
      tensorflow::grappler::MeasuringCostEstimator::PredictCosts(tensorflow::GraphDef const&, tensorflow::CostGraphDef*, tensorflow::grappler::Costs*) const in libtensorflow-core.a(measuring_cost_estimator.o)
      ...
  "nsync::nsync_from_time_point_(std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l> > >)", referenced from:
      std::__1::cv_status tensorflow::condition_variable::wait_for<long long, std::__1::ratio<1l, 1000l> >(tensorflow::mutex_lock&, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000l> >) in libtensorflow-core.a(allocator_retry.o)
      std::__1::cv_status tensorflow::condition_variable::wait_for<long long, std::__1::ratio<1l, 1000000l> >(tensorflow::mutex_lock&, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l> >) in libtensorflow-core.a(direct_session.o)
      std::__1::cv_status tensorflow::condition_variable::wait_for<long long, std::__1::ratio<1l, 1000000l> >(tensorflow::mutex_lock&, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l> >) in libtensorflow-core.a(rendezvous.o)
      std::__1::cv_status tensorflow::condition_variable::wait_for<long long, std::__1::ratio<1l, 1000000l> >(tensorflow::mutex_lock&, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l> >) in libtensorflow-core.a(utils.o)
  "nsync::nsync_mu_unlock(nsync::nsync_mu_s_*)", referenced from:
      tensorflow::AllocatorRetry::AllocateRaw(std::__1::function<void* (unsigned long, unsigned long, bool)>, int, unsigned long, unsigned long) in libtensorflow-core.a(allocator_retry.o)
      tensorflow::BFCAllocator::AllocateRawInternal(unsigned long, unsigned long, bool) in libtensorflow-core.a(bfc_allocator.o)
      tensorflow::BFCAllocator::DeallocateRaw(void*) in libtensorflow-core.a(bfc_allocator.o)
      tensorflow::BFCAllocator::DeallocateRawInternal(void*) in libtensorflow-core.a(bfc_allocator.o)
      tensorflow::BFCAllocator::AddAllocVisitor(std::__1::function<void (void*, unsigned long)>) in libtensorflow-core.a(bfc_allocator.o)
      tensorflow::BFCAllocator::RequestedSize(void*) in libtensorflow-core.a(bfc_allocator.o)
      tensorflow::BFCAllocator::AllocatedSize(void*) in libtensorflow-core.a(bfc_allocator.o)
      ...
  "nsync::nsync_cv_wait_with_deadline(nsync::nsync_cv_s_*, nsync::nsync_mu_s_*, timespec, nsync::nsync_note_s_*)", referenced from:
      std::__1::cv_status tensorflow::condition_variable::wait_for<long long, std::__1::ratio<1l, 1000l> >(tensorflow::mutex_lock&, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000l> >) in libtensorflow-core.a(allocator_retry.o)
      std::__1::cv_status tensorflow::condition_variable::wait_for<long long, std::__1::ratio<1l, 1000000l> >(tensorflow::mutex_lock&, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l> >) in libtensorflow-core.a(direct_session.o)
      std::__1::cv_status tensorflow::condition_variable::wait_for<long long, std::__1::ratio<1l, 1000000l> >(tensorflow::mutex_lock&, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l> >) in libtensorflow-core.a(rendezvous.o)
      std::__1::cv_status tensorflow::condition_variable::wait_for<long long, std::__1::ratio<1l, 1000000l> >(tensorflow::mutex_lock&, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l> >) in libtensorflow-core.a(utils.o)
  "nsync::nsync_mu_lock(nsync::nsync_mu_s_*)", referenced from:
      tensorflow::AllocatorRetry::AllocateRaw(std::__1::function<void* (unsigned long, unsigned long, bool)>, int, unsigned long, unsigned long) in libtensorflow-core.a(allocator_retry.o)
      tensorflow::BFCAllocator::AllocateRawInternal(unsigned long, unsigned long, bool) in libtensorflow-core.a(bfc_allocator.o)
      tensorflow::BFCAllocator::DeallocateRaw(void*) in libtensorflow-core.a(bfc_allocator.o)
      tensorflow::BFCAllocator::DeallocateRawInternal(void*) in libtensorflow-core.a(bfc_allocator.o)
      tensorflow::BFCAllocator::AddAllocVisitor(std::__1::function<void (void*, unsigned long)>) in libtensorflow-core.a(bfc_allocator.o)
      tensorflow::BFCAllocator::RequestedSize(void*) in libtensorflow-core.a(bfc_allocator.o)
      tensorflow::BFCAllocator::AllocatedSize(void*) in libtensorflow-core.a(bfc_allocator.o)
      ...
  "nsync::nsync_cv_broadcast(nsync::nsync_cv_s_*)", referenced from:
      tensorflow::BFCAllocator::DeallocateRaw(void*) in libtensorflow-core.a(bfc_allocator.o)
      tensorflow::DirectSession::Run(tensorflow::RunOptions const&, std::__1::vector<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, tensorflow::Tensor>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, tensorflow::Tensor> > > const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::vector<tensorflow::Tensor, std::__1::allocator<tensorflow::Tensor> >*, tensorflow::RunMetadata*) in libtensorflow-core.a(direct_session.o)
      tensorflow::Notification::Notify() in libtensorflow-core.a(direct_session.o)
      tensorflow::Notification::Notify() in libtensorflow-core.a(graph_runner.o)
      tensorflow::Notification::Notify() in libtensorflow-core.a(cancellation.o)
      tensorflow::Notification::Notify() in libtensorflow-core.a(op_kernel.o)
      tensorflow::Notification::Notify() in libtensorflow-core.a(reader_base.o)
      ...
  "nsync::nsync_mu_init(nsync::nsync_mu_s_*)", referenced from:
      tensorflow::AllocatorRetry::AllocatorRetry() in libtensorflow-core.a(allocator_retry.o)
      tensorflow::AllocatorRetry::AllocatorRetry() in libtensorflow-core.a(allocator_retry.o)
      tensorflow::BFCAllocator::BFCAllocator(tensorflow::SubAllocator*, unsigned long, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in libtensorflow-core.a(bfc_allocator.o)
      tensorflow::(anonymous namespace)::get_device_factory_lock() in libtensorflow-core.a(device_factory.o)
      tensorflow::DirectSession::DirectSession(tensorflow::SessionOptions const&, tensorflow::DeviceMgr const*, tensorflow::DirectSessionFactory*) in libtensorflow-core.a(direct_session.o)
      tensorflow::DirectSession::Run(tensorflow::RunOptions const&, std::__1::vector<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, tensorflow::Tensor>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, tensorflow::Tensor> > > const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::vector<tensorflow::Tensor, std::__1::allocator<tensorflow::Tensor> >*, tensorflow::RunMetadata*) in libtensorflow-core.a(direct_session.o)
      tensorflow::DirectSession::PRunSetup(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*) in libtensorflow-core.a(direct_session.o)
      ...
  "nsync::nsync_cv_init(nsync::nsync_cv_s_*)", referenced from:
      tensorflow::AllocatorRetry::AllocatorRetry() in libtensorflow-core.a(allocator_retry.o)
      tensorflow::AllocatorRetry::AllocatorRetry() in libtensorflow-core.a(allocator_retry.o)
      tensorflow::DirectSession::RunState::RunState(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, long long, std::__1::vector<tensorflow::Device*, std::__1::allocator<tensorflow::Device*> > const*) in libtensorflow-core.a(direct_session.o)
      tensorflow::Executor::Run(tensorflow::Executor::Args const&) in libtensorflow-core.a(graph_runner.o)
      tensorflow::CancellationManager::CancellationManager() in libtensorflow-core.a(cancellation.o)
      tensorflow::AsyncOpKernel::Compute(tensorflow::OpKernelContext*) in libtensorflow-core.a(op_kernel.o)
      tensorflow::ReaderBase::GetNextWorkLocked(tensorflow::QueueInterface*, tensorflow::OpKernelContext*) const in libtensorflow-core.a(reader_base.o)
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Same case occurred in Android when linking with Tensorflow static lib.

However, the build script is cleaning the built nsync lib across different builds, and the nsync lib is stored in a place that is different than other dependency libs (e.g. protobuf).

So Here is to copy the generated libs into gen folder like other lib would be placed when generated, so user or `make` can manage them in a consist way.

Hope these make sense to you, thanks.

File: tensorflow/contrib/makefile/compile_nsync.sh (hunk 1 above)
Comment (resec) [reply tensorflow/contrib/makefile/compile_nsync.sh]: I am not intended add the `-p` option to the command, let me remove it, thanks for reminding.

**Review by resec — COMMENTED** [review]
please help to review again, thanks

**Comment by drpngx** [comment]
@resec just to understand the scenario better, why are you running two builds? What are the builds that you are running? thanks.

**Comment by resec** [comment]
@drpngx I have multiple projects are using different tf built model on Android/iOS mobile device. And I am pre-building the tf and its dependencies lib, before I build my projects(to save compilation time and unifying the tf version for my projects). It was done by gathering(copying) required generated libs and headers files from tf and its dependencies to my projects. 

Additional information, my projects are not building with bezel.

**Comment by martinwicke** [comment]
@drpngx do you have an update for this?

**Comment by drpngx** [comment]
Sorry, not sure why this was not sent: why can't use the `$arch` file that's already built, ie why do we need to copy this somewhere else?

**Comment by resec** [comment]
@drpngx 

> However, the build script is cleaning the built nsync lib across different builds, and the nsync lib is stored in a place that is different than other dependency libs (e.g. protobuf).
So Here is to copy the generated libs into gen folder like other lib would be placed when generated, so user or make can manage them in a consist way.

**Comment by resec** [comment]
@drpngx then users can not reuse those build scripts existed in tensorflow, and have to write their own one.

Any first-time user trying to build the tensorflow lib, would very likely meet I met, and feel strange and uncomfortable. User can either just like you said, write another script,  then "run the script, collect the artifact, before you start another build?", or like what I did, just add some lines somewhere to copy the libs.

Copying will remove those obtascles.

**Comment by martinwicke** [comment]
drpngx@ I have no objections to this, but I also skeptical about the long term prospects of this makefile build in general, so I may not be the best person to ask.

@petewarden Do you have an opinion on this?

**Comment by drpngx** [comment]
@resec the build drops the lib in a place that's specific to the arch etc. If you move it somewhere else, it loses the information. If it survives cleanups, then it's adding a confusing side effect with weaker guarantees on which people would accidentally depend. I'd rather we avoid this.

**Comment by resec** [comment]
@drpngx just FYI, the arch etc info is kept via folder/path just like tensorflow and protobuf lib.

**Comment by resec** [comment]
@drpngx please see 
https://github.com/resec/tensorflow/blob/d3784bd37d493faee266ee9311a5a4b91bb5d7e4/tensorflow/contrib/makefile/compile_nsync.sh#L133 
and 
https://github.com/resec/tensorflow/blob/d3784bd37d493faee266ee9311a5a4b91bb5d7e4/tensorflow/contrib/makefile/compile_nsync.sh#L191
The final destination folder is arch-related.

**Comment by drpngx** [comment]
Why is it sometimes with arch (as in the two examples) and sometimes not (e.g. mac)?

**Comment by resec** [comment]
@drpngx it is simply following the current handling of tensorflow, protobuf lib.

If the lib is building for mobile, then lib is stored in a arch-related folder, if is it building for "server-side" (I guess the designer somehow assumed that if it is building for linux/mac, then it is building for host), then the lib is just stored into a "lib" without arch info.

If we want to change them, I think we should change them (including tensorflow, etc.) all together.

**Comment by drpngx** [comment]
Isn't what `$bin_prefix` is doing?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Pull request closed without merging**